sdks/java/core/src/main/java/org/apache/beam/sdk/io/OffsetBasedSource.java

```diff
@@ -315,6 +330,30 @@ public Double getFractionConsumed() {
     }
 
     @Override
+    public long getParallelismConsumed() {
+      return rangeTracker.getSplitPointsProcessed();
+    }
+
+    @Override
+    public long getParallelismRemaining() {
+      if (isDone()) {
+        return 0;
+      }
+      if (!isStarted()) {
+        // Note that even if the current source does not allow splitting, we don't know that
+        // it's non-empty so we return UNKNOWN instead of 1.
+        return BoundedReader.PARALLELISM_UNKNOWN;
+      }
+      if (!getCurrentSource().allowsDynamicSplitting()) {
+        return 1;
+      }
+      if (getCurrentOffset() + 1 >= rangeTracker.getStopPosition()) {
+        return 1;
```

Comment:
This case is confusing. I think it means that the current position is past the stop positino -- shouldn't that mean we have no more parallelism? Add a comment?